tests/test_moisture_sensor.py

```diff
@@ -7,21 +7,38 @@
 
 class MoistureSensorTest(unittest.TestCase):
 
-    def test_moisture_50(self):
+    def test_moisture_read_succeeds_when_adc_succeeds(self):
         """Moisture level read should be the average of the two readings."""
         mock_adc = mock.Mock()
         mock_adc.read_adc.side_effect = [300, (1023 - 400)]
         mock_pi_io = mock.Mock()
         mock_clock = mock.Mock()
         channel = 1
-        gpio_pins = (12, 16)
-        sensor = moisture_sensor.MoistureSensor(mock_adc, mock_pi_io, channel,
-                                                gpio_pins, mock_clock)
-        moisture_level = sensor.moisture()
-
-        self.assertIn(mock.call(12), mock_pi_io.turn_pin_on.call_args_list)
-        self.assertIn(mock.call(12), mock_pi_io.turn_pin_off.call_args_list)
-        self.assertIn(mock.call(16), mock_pi_io.turn_pin_on.call_args_list)
-        self.assertIn(mock.call(16), mock_pi_io.turn_pin_off.call_args_list)
-        mock_clock.wait.assert_called_once()
-        self.assertEqual(moisture_level, 350)
+        gpio_pin_1 = 12
+        gpio_pin_2 = 16
+        sensor = moisture_sensor.MoistureSensor(
+            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)
+
+        self.assertEqual(350, sensor.moisture())
+        self.assertEqual([mock.call(12), mock.call(16)],
+                         mock_pi_io.turn_pin_on.call_args_list)
+        self.assertEqual(
+            [mock.call(16), mock.call(12), mock.call(12), mock.call(16)],
+            mock_pi_io.turn_pin_off.call_args_list)
+        mock_clock.wait.assert_called_once_with(0.1)
+
+    def test_moisture_sensor_pins_turned_off_if_adc_raises_exception(self):
+        mock_adc = mock.Mock()
+        mock_adc.read_adc.side_effect = ValueError()
+        mock_pi_io = mock.Mock()
+        mock_clock = mock.Mock()
+        channel = 1
+        gpio_pin_1 = 12
+        gpio_pin_2 = 16
+        sensor = moisture_sensor.MoistureSensor(
+            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)
+
+        with self.assertRaises(ValueError):
+            sensor.moisture()
+            self.assertEqual([mock.call(16), mock.call(12), mock.call(16)],
+                             mock_pi_io.turn_pin_off.call_args_list)
```

Comment:
(optional) I'd like it to be more explicit that we're testing that the pins be left in an "off" state. The way we're currently testing is kind of roundabout because we're verifying that a particular sequence was called, but it's possible for the implementation to call that sequence and still leave the pins in an "on" state.

Here's an idea, but I'm not 100% happy with it because it really should be something we have in both tests (because it should leave pins in off state no matter what) but it's kind of a large amount of code to copy/paste both places, but I don't have any good ideas about how to share the code between both tests.

```python
    def test_moisture_sensor_pins_turned_off_if_adc_raises_exception(self):
        mock_adc = mock.Mock()
        mock_adc.read_adc.side_effect = ValueError()
        pin_state = collections.defaultdict(lambda: False)

        def mock_turn_pin_on(pin):
            pin_state[pin] = True

        def mock_turn_pin_off(pin):
            pin_state[pin] = False

        mock_pi_io = mock.Mock()
        mock_pi_io.turn_pin_on.side_effect = mock_turn_pin_on
        mock_pi_io.turn_pin_off.side_effect = mock_turn_pin_off
        mock_clock = mock.Mock()
        channel = 1
        gpio_pin_1 = 12
        gpio_pin_2 = 16
        sensor = moisture_sensor.MoistureSensor(
            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)

        with self.assertRaises(ValueError):
            sensor.moisture()

        # Make sure both pins are left in off state.
        self.assertFalse(pin_state[12])
        self.assertFalse(pin_state[16])
```